go/vt/vtgate/planbuilder/operators/projection_pushing.go

```diff
@@ -192,6 +192,22 @@ func pushProjectionToOuterContainer(ctx *plancontext.PlanningContext, p *Project
 	return src, Rewrote("push projection into outer side of subquery container")
 }
 
+// nullInNullOutExpr returns true if the expression will return NULL if any of its inputs are NULL
+// When we are evaluating an ApplyJoin, the expressions that have any dependency on the outer side of the join
+// will be sent to the outer side of the join. If the expression is null intolerant, then we can push it down,
+// and the result would be NULL for missing matches from the outer side. If the expression is something that can
+// return values other than NULL, like `COALESCE(tbl.foo, 'bar')`, then we can't push it down, because we would
+// get a different result if the outer side is missing.
+func nullInNullOutExpr(expr sqlparser.Expr) bool {
+	// TODO: This is a very basic implementation. We should expand this to handle more cases.
+	switch expr.(type) {
+	case *sqlparser.ColName:
+		return true
+	default:
+		return false
+	}
+}
```

Comment:
Shouldn't we check for all cases where this function should return true i.e. `IFNULL` before merging this PR?

Reply:
Returning false here means that we do not allow the expression to be pushed under the `ApplyJoin`, which is the behaviour we have today. We don't need to add `IFNULL` and `COALESCE` here, they are already take care of with the `default` case. We do need to add expressions that will return `NULL` if their inputs are `NULL` - such as `+`, `Lower()`, etc...

Reply:
That makes more sense, thank you for explaining.

Reply:
We should move this to the AST package.